### PR TITLE
Parallel

### PR DESCRIPTION
**Pull request by sbaughman**
Changes initial document spellcheck to be split into block-level elements so they can occur in parallel.  Also fixes a bug that caused changes to text inside a spellcheck span to trigger a whole-document spellcheck

### CODE REVIEW

File: plugins/nanospell/plugin.js [hunk 1]
```diff
@@ -429,23 +416,36 @@
 				triggerSpelling((spellFastAfterSpacebar && (ch8r === CHARCODES.SPACE || ch8r === CHARCODES.LF || ch8r === CHARCODES.CR)), target)
 			}
 
+			function findNearestParentBlock(element) {
+				var elementPath = new CKEDITOR.dom.elementPath(element),
+					elements = elementPath.elements,
+					i;
+
+				for (i in elements) {
```
Comment (caffodian) [inline plugins/nanospell/plugin.js]: Is this actually consistent?  

Memory is foggy, but elements is an array, right?  

See https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Statements/for...in#Array_iteration_and_for...in

Comment (caffodian) [inline plugins/nanospell/plugin.js]: Worried that this is actually doing "findRandomBlockInPath" in some browser, somewhere.

File: plugins/nanospell/plugin.js [hunk 2]
```diff
@@ -358,37 +355,26 @@
 			}
 
 			function checkNow(rootElement) {
-				if (!selectionCollapsed() || self._spellCheckInProgress) {
+				rootElement = rootElement || editor.editable();
+
+				if (!selectionCollapsed() || spellCheckInProgress(rootElement)) {
```
Comment (caffodian) [inline plugins/nanospell/plugin.js:361]: Much nicer!

I'm wondering if `triggerSpelling` should be replaced with this `checkNow` method,

they do really similar things, but this seems to be doing more checks now.  Maybe not good to roll into this PR.

File: tests/smoke/events.js [hunk 3]
```diff
@@ -93,7 +95,7 @@
 				var secondParagraph = editor.editable().getChild(1);
 
 				// no ajax call required on the second run, since words are repeats.
-				observer.assert(["spellCheckComplete", "startMarkTypos", "startScanWords"]);
+				observer.assert(["spellCheckComplete", "startRender", "startScanWords"]);
 				observer.assertRootIs(secondParagraph);
 			}
```
Comment (caffodian) [inline tests/smoke/events.js:100]: Do you think it'd be worth adding a test case just to make sure the new logic for blocking duplicate spellchecks?

I'm thinking a case like

```
<ol><li>something
<ul><li>somethingnested</li></li></ul>
</li></ol>
```

If my understanding is correct, spellchecking "something" will cause "somethingnested" to be skipped

File: plugins/nanospell/plugin.js [hunk 4]
```diff
@@ -421,6 +407,7 @@
 				var spellCheckSpan = elementPath.contains(isSpellCheckSpan);
 
 				if (spellCheckSpan) {
+					target = findNearestParentBlock(target);
```
Comment (caffodian) [inline plugins/nanospell/plugin.js:412]: Ah.  I was going to ask about this.

I thought more about my earlier suggestion to fold `checkNow()` and `triggerSpelling()` together.  I was thinking we could enforce that a block was passed in, so as to avoid the problem you asked about in hipchat.  Then I see that you are enforcing it down here.  :)

**Comment by caffodian** [comment]
Looks good so far.  

One idea for how to manual test this:
- copy this new version into site_media/lib/ckeditor-common/nanospell on the PolicyStat dev box
- override an existing policy.  Replace it with a LOT of generated lorem ipsum text (you may need to runserver with the following env due to a unicode issue w/ django debug toolbar `PSTAT_HIDE_DEBUG_TOOLBAR=1 ./pstat/manage.py runserver 0.0.0.0:8000`
- save it
- open the policy again and see if it's more responsive on the initial load.

**Comment by caffodian** [comment]
I'd also like to work through that nested list case again.  Another similar case would be for paragraph tags inside table cells, which ckeditor also gets weird with, as far as `getNextParagraph` is concerned.

**Comment by caffodian** [comment]
I really appreciate that you looked at it with fresh eyes and were able to make some clarity improvements as well.  I had been deep into the rewrite about a month ago and lost a lot of that context.

File: plugins/nanospell/plugin.js (hunk 1 above)
Comment (sbaughman) [reply plugins/nanospell/plugin.js]: Yep, you're right.  I was basing this off of a loop in one of the functions in the original plugin, but now I realize that order was not important in that case.  There's another one of these somewhere in here.  I'll fix them both

File: tests/smoke/events.js (hunk 3 above)
Comment (caffodian) [reply tests/smoke/events.js:100]: For our own sanity, this might be a good time to have a test helper that clears `suggestions` and `suggestionscache`, which we could run in the teardown message.  Then we don't need to keep creating new variations of 'foo' and 'asdf' for new tests.

File: plugins/nanospell/plugin.js (hunk 1 above)
Comment (caffodian) [reply plugins/nanospell/plugin.js]: In Python, for...in does respect order, which means I do this a lot :)

File: tests/smoke/events.js (hunk 3 above)
Comment (sbaughman) [reply tests/smoke/events.js:100]: Absolutely.

**Comment by caffodian** [comment]
This is definitely a later thing, unless the cause is really obvious.  

You can type (including inserting new elements) while spellcheck is running.

however, spellcheck does not trigger.  Had to type more into the block after it was done to trigger something.

**Comment by caffodian** [comment]
An actual bug - things like the source dialog don't work while spellcheck is running.  This is probably due to some weird race condition but eventually resolves itself when all the spans have been marked.   

Might have something to do with the filters, which remove spans before sending it off to the source dialog

File: tests/smoke/events.js [hunk 5]
```diff
@@ -105,6 +107,54 @@
...
+
+				// wait for initial spellcheck to complete
+				wait();
+
+				function completeFirstSpellcheck() {
+
+					observer.assert(["spellCheckComplete", "startRender", "startCheckWordsAjax", "spellCheckComplete", "startRender", "startCheckWordsAjax", "startScanWords"]);
+
+					// set outer li to show that a spellcheck is in progress
+					outer.setCustomData('spellCheckInProgress', true);
```
Comment (caffodian) [inline tests/smoke/events.js:136]: Ah.  This is basically faking an in progress spellcheck, since of course, we don't have an event to be able to pause on the spellcheck on the outer list.  Is my understanding correct?

File: tests/smoke/events.js [hunk 6]
```diff
@@ -145,7 +145,8 @@
 				i,
 				root;
 
-			for (i=0, event=events[i]; i<events.length; i++) {
+			for (i=0; i<events.length; i++) {
```
Comment (caffodian) [inline tests/smoke/events.js:197]: doh.  pretty sure I initially was trying to be clever and just ended up just not iterating.  Good catch that makes sense now.

**Review by caffodian — APPROVED** [review]
I'm pretty satisfied with the state of things as is.  I've documented a few things outside of this "review" that we can investigate as future work.  I can probably get the beta testing for putting this into PolicyStat done tomorrow.  If my initial testing results hold in other browsers, I think we can at least get this PR rolled out to actual users soon!

File: tests/smoke/events.js (hunk 5 above)
Comment (sbaughman) [reply tests/smoke/events.js:136]: Yep.  I just let the initial check complete, then reset the outer block data manually to reflect a check in progress
